### ISSUE TITLE
java.util.Collection.parallelStream raises SecurityException when called inside controller's action

### ISSUE THREAD

**Issue by fabgilson** [playframework/playframework#7751]
### JDK Oracle 1.8.0_144

Usage of Collection.parallelStream() throws java.lang.SecurityException: setContextClassLoader when used inside controller's action scope. Was fully working with Play Java version 2.5.15.

### Workaround

using stream() fixes the problem.

### Full stack

```
java.lang.SecurityException: setContextClassLoader
	at java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:273)
	at java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:280)
	at java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1592)
	at play.core.j.HttpExecutionContext$$anon$2.run(HttpExecutionContext.scala:56)
	at akka.dispatch.TaskInvocation.run(AbstractDispatcher.scala:40)
	at akka.dispatch.ForkJoinExecutorConfigurator$AkkaForkJoinTask.exec(ForkJoinExecutorConfigurator.scala:43)
	at akka.dispatch.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
	at akka.dispatch.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
	at akka.dispatch.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
	at akka.dispatch.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)
Caused by: java.lang.SecurityException: java.lang.SecurityException: setContextClassLoader
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
	at java.util.concurrent.ForkJoinTask.getThrowableException(ForkJoinTask.java:593)
	at java.util.concurrent.ForkJoinTask.reportException(ForkJoinTask.java:677)
	at java.util.concurrent.ForkJoinTask.invoke(ForkJoinTask.java:735)
	at java.util.stream.ForEachOps$ForEachOp.evaluateParallel(ForEachOps.java:160)
	at java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateParallel(ForEachOps.java:174)
	at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:233)
	at java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:418)
	at be.webdeb.presentation.web.controllers.viz.argument.ArgumentVizHolder.lambda$getLinks$10(ArgumentVizHolder.java:225)
	at java.util.HashMap.forEach(HashMap.java:1288)
	at be.webdeb.presentation.web.controllers.viz.argument.ArgumentVizHolder.getLinks(ArgumentVizHolder.java:222)
	at be.webdeb.presentation.web.controllers.viz.argument.ArgumentVizHolder.getAmountByShade(ArgumentVizHolder.java:357)
	at be.webdeb.presentation.web.views.html.viz.argument.argumentCartography$.apply(argumentCartography.template.scala:59)
	at be.webdeb.presentation.web.views.html.viz.argument.argumentVisualization$.apply(argumentVisualization.template.scala:225)
	at be.webdeb.presentation.web.views.html.viz.argument.argumentVisualization$.render(argumentVisualization.template.scala:258)
	at be.webdeb.presentation.web.views.html.viz.argument.argumentVisualization.render(argumentVisualization.template.scala)
	at be.webdeb.presentation.web.controllers.viz.VizActions.lambda$argument$1(VizActions.java:153)
	at java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1590)
	... 7 common frames omitted
Caused by: java.lang.SecurityException: setContextClassLoader
	at java.util.concurrent.ForkJoinWorkerThread$InnocuousForkJoinWorkerThread.setContextClassLoader(ForkJoinWorkerThread.java:244)
	at play.api.inject.ContextClassLoaderInjector.withContext(Injector.scala:125)
	at play.api.inject.ContextClassLoaderInjector.instanceOf(Injector.scala:118)
	at be.webdeb.presentation.web.controllers.entry.ContributionHolder.<init>(ContributionHolder.java:57)
	at be.webdeb.presentation.web.controllers.entry.argument.LinkHolder.<init>(LinkHolder.java:82)
	at be.webdeb.presentation.web.controllers.viz.argument.LinkVizHolder.<init>(LinkVizHolder.java:54)
	at be.webdeb.presentation.web.controllers.viz.argument.ArgumentVizHolder.computeCartography(ArgumentVizHolder.java:629)
	at be.webdeb.presentation.web.controllers.viz.argument.ArgumentVizHolder.lambda$null$9(ArgumentVizHolder.java:232)
	at java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:184)
	at java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:175)
	at java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1374)
	at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:481)
	at java.util.stream.ForEachOps$ForEachTask.compute(ForEachOps.java:291)
	at java.util.concurrent.CountedCompleter.exec(CountedCompleter.java:731)
	at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:289)
	at java.util.concurrent.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1056)
	at java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1692)
	at java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:157)
```

**Comment by marcospereira** [comment]
Probably related to https://github.com/playframework/playframework/pull/7552 just like https://github.com/playframework/playframework/issues/7615. Could you please provide a complete example that reproduces the problem?

Thanks.

**Comment by fgi-github** [comment]
After some digging, this is the @Inject inside a class where I explicitly use the Play.current().injector().instanceOf() to get an injected class that is called inside a parallelStream context. Hard to provide a complete MWE example on this. But definitely the same troubles as in #7615.

**Comment by richdougherty** [comment]
> Hard to provide a complete MWE example on this.

Any info you can provide, e.g. specific `parallelStream` usage, thread pool config, can be helpful to use in debugging this one.

The solution for #7624 was to be more explicit about which thread pool to use when using the `CompletionStage` API. I did some searching around and it looks like the underlying problem is that many Java APIs use a default fork-join pool (`ForkJoinPool.commonPool`) which has very restrictive thread security (see `ForkJoinWorkerThread$InnocuousForkJoinWorkerThread`).

There's unfortunately no way to override the common pool by default, but it looks like you *can* get a Java API to use your own fork-join pool if you call it from within one, e.g. Example from https://stackoverflow.com/questions/21163108/custom-thread-pool-in-java-8-parallel-stream/22269778#22269778:

```java
ForkJoinPool forkJoinPool = new ForkJoinPool(2);
forkJoinPool.submit(() ->
    //parallel task here, for example
    IntStream.range(1, 1_000_000).parallel().filter(PrimesPrint::isPrime).collect(toList())
).get();
```

Can you try doing something like this, just to test if the concept works? I think this will work as it is, but if you still get an exception then you may need to change `new ForkJoinPool(2)` to use a constructor that supplies your own `ForkJoinWorkerThreadFactory`.

I'll be interested to hear if that works for you. If it works then we can look into how to change Play's own fork-join pool so that it is used instead of Java's common and restrictive fork-join pool. If we can do that then methods like `parallelStream` should work automatically.

**Comment by fgi-github** [comment]
Nice catch, it works by manually creating the pool as you suggest, just creating a ForkJoinPool. Many thanks!

So, to explain more on the "workflow" in my controller's action method :
1. I receive a GET call to retrieve some data from DB (no problem there)
2. map those DB objects into presentation objects into a parallelStream().map().collect()
3. those objects have "manual" injected fields retrieved by play.api.Play.current().injector().instanceOf()
4. a java.lang.SecurityException: setContextClassLoader is raised from the injector().instanceOf() line when instantiating the object.

Hope this "trace" helps you.

**Comment by fgi-github** [comment]
Some more info on this. The same SecurityException arises when calling an external web service (REST) and chaining with the creation of objects as described upper (with calling the injector().instanceOf() manually for fields). By tracing the execution threads, I found out that:
1. GET call (in controller) is executed into the "application-akka.actor.default-dispatcher-X" thread
2. the REST call is managed as a CompletionStage into a dedicated class using a play.libs.ws.WSClient. It is also run into the same "application-akka.actor.default-dispatcher-X" thread
3. chaining the result (with a thenApply for example) make the execution moving to a "ForkJoinPool.commonPool-worker-x" and not in the "application-akka" pool.

Could not find a complete workaround to this.

**Comment by marcospereira** [comment]
> chaining the result (with a thenApply for example) make the execution moving to a "ForkJoinPool.commonPool-worker-x" and not in the "application-akka" pool.

You can be explicit about the `Executor` being used by `thenApply` too, but by using `thenApplyAsync` like:

```java
import play.mvc.Controller;
import scala.concurrent.ExecutionContextExecutor;

import javax.inject.Inject;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionStage;

public class FooController extends Controller {

    private final ExecutionContextExecutor ec;

    @Inject
    public Foo(ExecutionContextExecutor ec) {
        this.ec = ec;
    }

    public void bar() {
        someCompletionStage().thenApplyAsync(String::valueOf, ec);
    }
}
```

This is using the default execution context and you may want to avoid it if you are doing block operations inside `thenApplyAsyn`. If you are using Play 2.6, you can inject a [custom execution contexts](https://www.playframework.com/documentation/2.6.x/Highlights26#Defining-a-CustomExecutionContext-in-Java) or you can create an `Executor` by yourself. The important aspect here is to avoid mixing `ForkJoinPool.commonPool()` (used by parallel streams and as the default executor when manipulating `CompletionStages`) **and** Play injector at the same time.

**Comment by fgi-github** [comment]
ok, thanks. Missed that part of the documentation of CompletionStage-related features.

**Comment by richdougherty** [comment]
Thanks @fgi-github for the extra info!

Note to self (or whoever fixes this)…

From https://github.com/google/guice/wiki/ClassLoading it seems that Guice uses a ClassLoader descended from the target class's ClassLoader. In our case often the target classes will be built-in Play classes, and thus on the non-reloading ClassLoader, but the implementations to load may be the user's application classes, and thus on the reloading ClassLoader. I don't see a way to override the Guice ClassLoader, but hopefully there is a workaround somewhere…

An important distinction to make is that once the instance or provider Class has been *loaded*, its constructor or `get` method will be *called* in the thread context loader. I was testing the calling part here: https://github.com/playframework/playframework/pull/7552/files#diff-bbe0c486aa559dd67bda2bb537986393R49 to fix #7535.

So when `injector.instanceOf[PlayClass]` tries to instantiate `AppClass`:
1. loads `AppClass` using `PlayClass.getClass.getClassLoader`
2. instantiates `AppClass` in the current thread (implicitly using `Thread.currentThread.getContextClassLoader`)

Problem 2 is what #7552 fixes. The problem in this thread is that the fix for problem 2 (setting the context ClassLoader) prevents Guice from working in the default commonPool FJP. Whoever fixes this should update `GuiceInjectorBuilderSpec` and make sure that Guice works with parallel stream operations, but we don't want to regress the fix on problem 2 while we do it.

**Comment by mkurz** [comment]
Read [my comment here](https://github.com/playframework/playframework/issues/8167#issuecomment-490645264) why I belive this issue will be fixed with Play 2.8.

**Comment by mkurz** [comment]
I finally found time to write a test to confirm `SecurityException` will not be thrown anymore when using `map`, `filter`, etc. on a **parallel** stream within a Play action method. See #9594

However, when using `CompletableFuture.supplyAsync(...)` and `thenxxxAsync` methods you still have to pass an explicit ec, because Java uses the common thread pool for this method by default. There is nothing Play can do about this (it's no an issue with Play actually).

I suggest closing this issue as soon as #9594 got merged.
